### PR TITLE
W3 - phongle - Create save data screen

### PR DESCRIPTION
**Pull request by at-phongle**
#### Related issues (Optional)
Task #5

#### Description (Required)
- [x] SharePreference
- [x] ExternalStorage
- [x] Sqlite

#### Do bài trước trong package model e có tạo class user nên bài này e đổi tên class từ user sang class person để tránh conflix.

### CODE REVIEW

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/ExternalStorageActivity.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            Log.d("w", getResources().getString(R.string.edt_write_file_fail));
+        } catch (IOException e) {
+            Log.d("ww", getResources().getString(R.string.edt_write_file_fail));
+        }
+    }
+
+    private void readData() {
+        String sdcard = Environment.getExternalStorageDirectory().getAbsolutePath() + getResources().getString(R.string.file_name);
+        try {
+            InputStreamReader reader = new InputStreamReader(new FileInputStream(sdcard));
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/ExternalStorageActivity.java]: ERROR: [I18N][DM_DEFAULT_ENCODING] Found reliance on default encoding in vn.asiantech.internship.save_data.ExternalStorageActivity.readData(): new java.io.InputStreamReader(InputStream)
http://findbugs.sourceforge.net/bugDescriptions.html#DM_DEFAULT_ENCODING

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/ExternalStorageActivity.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    @Override
+    public void onClick(View v) {
+        writeData();
+        readData();
+    }
+
+    private void writeData() {
+        String sdcard = Environment.getExternalStorageDirectory().getAbsolutePath() + getResources().getString(R.string.file_name);
+        try {
+            OutputStreamWriter writer = new OutputStreamWriter(new FileOutputStream(sdcard));
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/ExternalStorageActivity.java]: ERROR: [I18N][DM_DEFAULT_ENCODING] Found reliance on default encoding in vn.asiantech.internship.save_data.ExternalStorageActivity.writeData(): new java.io.OutputStreamWriter(OutputStream)
http://findbugs.sourceforge.net/bugDescriptions.html#DM_DEFAULT_ENCODING

**Comment by at-phongle** [comment]
@at-kietvo please review this.thanks

File: app/src/at-phongle/AndroidManifest.xml [hunk 3]
```diff
@@ -18,6 +21,10 @@
         <activity android:name=".calculator.CalculatorActivity" />
         <activity android:name=".recyclerview.PostActivity" />
         <activity android:name=".login.LoginActivity" />
+        <activity android:name=".save_data.SharePreferenceActivity" />
```
Comment (at-kietvo) [inline app/src/at-phongle/AndroidManifest.xml]: Pls remove `_` in package name

File: app/src/at-phongle/res/layout/activity_external_storage.xml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:orientation="vertical"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:gravity="center"
+    android:padding="20dp">
+    <EditText
```
Comment (at-kietvo) [inline app/src/at-phongle/res/layout/activity_external_storage.xml]: Pls reformat code.

File: app/src/at-phongle/res/layout/activity_app.xml [hunk 5]
```diff
@@ -23,4 +23,25 @@
         android:layout_width="match_parent"
         android:layout_height="wrap_content"
         android:text="@string/btn_recycler_view_screen" />
+    <Button
```
Comment (at-kietvo) [inline app/src/at-phongle/res/layout/activity_app.xml]: Reformat code.

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/SharePreferenceActivity.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    private void initListener() {
+        mBtnClickSharePrefer.setOnClickListener(this);
+        mEdtInputSharePrefer.addTextChangedListener(this);
+    }
+
+    @Override
+    public void onClick(View v) {
+        readPreference();
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/SharePreferenceActivity.java]: I want to read Preference when the view inited, not by click.

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/SharePreferenceActivity.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    private void savePreference() {
+        SharedPreferences sharedPreferences = getSharedPreferences("SharePrefer", MODE_PRIVATE);
+        SharedPreferences.Editor editor = sharedPreferences.edit();
+        String text = mEdtInputSharePrefer.getText().toString();
+        editor.putString("input", text);
+        editor.apply();
+    }
+
+    private void readPreference() {
+        SharedPreferences sharedPreferences = getSharedPreferences("SharePrefer", MODE_PRIVATE);
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/SharePreferenceActivity.java]: Define Key in static String

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/SharePreferenceActivity.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        mEdtInputSharePrefer.setText(sharedPreferences.getString("input", "not input text"));
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/SharePreferenceActivity.java]: Define Key in static String

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonActivity.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_person);
+        initViews();
+        initData();
+        initAdapter();
+    }
+
+    private void initViews() {
+        mRecyclerViewPerson = findViewById(R.id.recyclerViewPerson);
+
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonActivity.java]: Remove this line.

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonActivity.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    private void initAdapter() {
+        PersonAdapter mPersonAdapter = new PersonAdapter(mPersonList, this);
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonActivity.java]: change variable name. Remove` if it's not the global variable.

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonActivity.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+            mDBManager.addEmployee(mPersonList.get(2), mCompanyList.get(0));
+        } else {
+            mPersonList = list;
+        }
+    }
+
+    @Override
+    public void onItemClick(int position) {
+        Intent intent = new Intent(this, DetailCompanyActivity.class);
+        Log.d("xx", String.valueOf(position));
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonActivity.java]: Remove this Log

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonAdapter.java [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+import android.view.ViewGroup;
+import android.widget.RelativeLayout;
+import android.widget.TextView;
+
+import java.util.List;
+
+import vn.asiantech.internship.R;
+import vn.asiantech.internship.models.Person;
+
+public class PersonAdapter extends RecyclerView.Adapter<PersonAdapter.PersonViewHolder> {
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonAdapter.java]: Add Javadoc.

File: app/src/at-phongle/java/vn/asiantech/internship/data/DBManager.java [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+        }
+        return company;
+    }
+
+    public List<Company> getCompanyByPersonId(int id) {
+        List<Company> companyList = new ArrayList<>();
+        SQLiteDatabase db = this.getReadableDatabase();
+        String sql = "SELECT c." + COLUMN_COMPANY_NAME + ", c." + COLUMN_COMPANY_SLOGAN + " FROM " +TABLE_COMPANY
+                + " AS c INNER JOIN " + TABLE_EMPLOYEE + " AS e ON e." + COLUMN_EMPLOYEE_ID_COMPANY + " = c." + COLUMN_COMPANY_ID
+                + " WHERE e." + COLUMN_EMPLOYEE_ID_PERSON + " =?";
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/data/DBManager.java]: This code ok, but why from `?` we have warning, pls check this and fix warning.

File: app/src/at-phongle/java/vn/asiantech/internship/data/DBManager.java [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+            company = new Company(id, cursor.getString(1), cursor.getString(2));
+            cursor.close();
+        }
+        return company;
+    }
+
+    public List<Company> getCompanyByPersonId(int id) {
+        List<Company> companyList = new ArrayList<>();
+        SQLiteDatabase db = this.getReadableDatabase();
+        String sql = "SELECT c." + COLUMN_COMPANY_NAME + ", c." + COLUMN_COMPANY_SLOGAN + " FROM " +TABLE_COMPANY
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/data/DBManager.java]: This code can work well without inner join.
You can use both from TABLE_EMPLOYEE,TABLE_COMPANY
and check condition in `where`

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonActivity.java [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+import vn.asiantech.internship.R;
+import vn.asiantech.internship.data.DBManager;
+import vn.asiantech.internship.models.Company;
+import vn.asiantech.internship.models.Person;
+
+public class PersonActivity extends AppCompatActivity implements PersonAdapter.OnItemClickListener {
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonActivity.java]: add JavaDoc

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonActivity.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+        } else {
+            mPersonList = list;
+        }
+    }
+
+    @Override
+    public void onItemClick(int position) {
+        Intent intent = new Intent(this, DetailCompanyActivity.class);
+        Log.d("xx", String.valueOf(position));
+        intent.putExtra("id", mPersonList.get(position).getId());
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonActivity.java]: Put key in constant (static String)

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonAdapter.java [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    public void onBindViewHolder(PersonAdapter.PersonViewHolder holder, int position) {
+        holder.onBindData(mPersonList, position);
+    }
+
+    @Override
+    public int getItemCount() {
+        return mPersonList.size();
+    }
+
+    static class PersonViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonAdapter.java]: Add Javadoc pls

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonAdapter.java [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+            mTvPersonName = itemView.findViewById(R.id.tvPersonName);
+            mTvAge = itemView.findViewById(R.id.tvAge);
+        }
+
+        private void initListener() {
+            mRlItemPerson.setOnClickListener(this);
+        }
+
+        private void onBindData(List<Person> personList, int postion) {
+            Person person = personList.get(postion);
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonAdapter.java]: Check null, pls

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonAdapter.java [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+
+        @Override
+        public void onClick(View v) {
+            if (mOnItemClickListener != null) {
+                mOnItemClickListener.onItemClick(getAdapterPosition());
+            }
+        }
+    }
+
+    public interface OnItemClickListener {
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/PersonAdapter.java]: Javadoc pls

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/DetailCompanyActivity.java [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+
+    private void initViews() {
+        mTvCompanyName = findViewById(R.id.tvCompanyName);
+        mTvCompanySlogan = findViewById(R.id.tvCompanySlogan);
+    }
+
+    private void setViews() {
+        Intent intent = getIntent();
+        mDBManager = new DBManager(this);
+        int id = intent.getIntExtra("id", 1);
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/DetailCompanyActivity.java]: Add Key to static string

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/DetailCompanyActivity.java [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+        Log.d("vv", String.valueOf(id));
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/DetailCompanyActivity.java]: Remove Log.

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/DetailCompanyActivity.java [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+    private TextView mTvCompanyName;
+    private TextView mTvCompanySlogan;
+    private List<Company> mCompanyList;
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_detail_company);
+        initViews();
+        setViews();
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/DetailCompanyActivity.java]: change method name to `updateViews`

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/DetailCompanyActivity.java [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+
+import vn.asiantech.internship.R;
+import vn.asiantech.internship.data.DBManager;
+import vn.asiantech.internship.models.Company;
+
+public class DetailCompanyActivity extends AppCompatActivity {
+    private DBManager mDBManager;
+    private TextView mTvCompanyName;
+    private TextView mTvCompanySlogan;
+    private List<Company> mCompanyList;
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/DetailCompanyActivity.java]: Convert to locally

File: app/src/at-phongle/java/vn/asiantech/internship/save_data/ExternalStorageActivity.java [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+
+import vn.asiantech.internship.R;
+
+
+public class ExternalStorageActivity extends AppCompatActivity implements View.OnClickListener {
```
Comment (at-kietvo) [inline app/src/at-phongle/java/vn/asiantech/internship/save_data/ExternalStorageActivity.java]: add Javadoc

File: app/src/at-phongle/java/vn/asiantech/internship/saveData/DetailCompanyActivity.java [hunk 25]
```diff
@@ -0,0 +1,47 @@
+package vn.asiantech.internship.saveData;
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/saveData/DetailCompanyActivity.java]: WARNING: [PackageCase] Package name contains upper case characters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#PackageCase

File: app/src/at-phongle/java/vn/asiantech/internship/saveData/ExternalStorageActivity.java [hunk 26]
```diff
@@ -0,0 +1,85 @@
+package vn.asiantech.internship.saveData;
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/saveData/ExternalStorageActivity.java]: WARNING: [PackageCase] Package name contains upper case characters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#PackageCase

File: app/src/at-phongle/java/vn/asiantech/internship/saveData/PersonActivity.java [hunk 27]
```diff
@@ -0,0 +1,80 @@
+package vn.asiantech.internship.saveData;
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/saveData/PersonActivity.java]: WARNING: [PackageCase] Package name contains upper case characters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#PackageCase

File: app/src/at-phongle/java/vn/asiantech/internship/saveData/PersonAdapter.java [hunk 28]
```diff
@@ -0,0 +1,87 @@
+package vn.asiantech.internship.saveData;
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/saveData/PersonAdapter.java]: WARNING: [PackageCase] Package name contains upper case characters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#PackageCase

File: app/src/at-phongle/java/vn/asiantech/internship/saveData/SharePreferenceActivity.java [hunk 29]
```diff
@@ -0,0 +1,58 @@
+package vn.asiantech.internship.saveData;
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/saveData/SharePreferenceActivity.java]: WARNING: [PackageCase] Package name contains upper case characters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#PackageCase

File: app/src/at-phongle/java/vn/asiantech/internship/saveData/SharePreferenceActivity.java [hunk 30]
```diff
@@ -0,0 +1,58 @@
...
+import vn.asiantech.internship.R;
+
+/**
+ * Created by phongle on 11/12/2560.
+ * SharePreferenceActivity
+ */
+public class SharePreferenceActivity extends AppCompatActivity implements View.OnClickListener {
+    private EditText mEdtInputSharePrefer;
+    private Button mBtnClickSharePrefer;
+    private static String SHARE_PREFERENCE = "SharePrefer";
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/saveData/SharePreferenceActivity.java]: WARNING: [SuspiciousConstantFieldName] The field name indicates a constant but its modifiers do not
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#SuspiciousConstantFieldName

File: app/src/at-phongle/java/vn/asiantech/internship/saveData/SharePreferenceActivity.java [hunk 31]
```diff
@@ -0,0 +1,58 @@
...
+    private static String SHARE_PREFERENCE_KEY = "input";
```
Comment (at-dev-support-01) [inline app/src/at-phongle/java/vn/asiantech/internship/saveData/SharePreferenceActivity.java]: WARNING: [SuspiciousConstantFieldName] The field name indicates a constant but its modifiers do not
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#SuspiciousConstantFieldName

**Comment by at-kietvo** [comment]
@at-phongle : pls check comment.s

**Comment by at-phongle** [comment]
@at-kietvo please review this. Thanks

**Pull request merged**